### DISCUSSION TITLE
How to use Named Entity Recognition in structure text like invoices?

### DISCUSSION THREAD

**Discussion by rsoika** [stanfordnlp/CoreNLP#1287] [General]
I have a question about how to best use CoreNLP for Names Entity Recognition missing natural language text blocks.

My goal is to extract named entities from an invoice document. As far as I understand, an invoice does not provide the unstructured plain text which is usually expected from NLP.

An typical text example for an invoice text analyzed with NLP ML which I found often in the Internet, looks like this:

> “Partial invoice (€100,000, so roughly 40%) for the consignment C27655 we shipped on 15th August to London from the Make Believe Town depot. INV2345 is for the balance.. Customer contact (Sigourney) says they will pay this on the usual credit terms (30 days).”

I understand that NLP loves this kind of text. But this text example is not typical for an invoice document. Text extracted form a Invoice PDF (using Apache Tika) usually looks more like this:

```
Client no: Invoice no: Invoice date: Due date:
1000011128 DEAXXXD220012269 26-Jul-2022 02-Aug-2022
Invoice to: Booking Reference
LOGISTCS GMBH Client Reference :
DEMOSTRASSE 2-6 Comments:
28195 BREMEN
Germany
Vessel : Voy : Place of Receipt : POL: B/LNo:
XXX JUBILEE NUBBBW SAV33NAH, GA ME000243
ETA: Final Destination : POD:
15-Jul-2022 ANTWERP, BELGIUM
Charge Quantity(days) x Rate Currency Total ROE Total EUR VAT
STORAGE_IMP_FOREIGN 1 day(s) x 30,00 EUR EUR 30,00 1,000000 30,00 0,00
```

Is NLP in general the wrong approach to train the recognition of meta data from an invoice document? 
Is there a way to setup up a pipeline in CoreNLP which is more useful for structured text in a invoice document?

Thanks for any tips

===
Ralph

**Comment by AngledLuffa** [comment]
Do these invoices have a consistent format?  This seems like something
you'd be better served by using regular expressions.  There is the
tokensregex package if you want to give that a try

**Reply by rsoika** [reply]
Hi @AngledLuffa, no the invoices do not have a consistent format. It is more a endless flow from incoming invoices from different suppliers. Of course some things are typical identical over all invoices like IBAN, BIC/SWIFT, or the VAT-ID. 
So using regular expressions maybe makes sense for such entities. And some suppliers send several time a invoice. So there are a some how recurring formats. 

What I do not understand in CoreNLP is that it looks like a training model provides only data line by line. But for an invoice - for example the total amount - is mostly somewhere in the last lines of the document. Is CoreNLP able to take respect of where in a document the entity is located?

What I mean is: an invoice can have a lot of separate amounts (many many numbers). Is CoreNLP able to learn that the Total is more at the end of a invoice document or is CoreNLP unable to see a difference:

```
Part Quantity  Rate Currency Total 
X1   5             1,50     EUR   7,50
X2  10             1,00     EUR   10,00
Total                       EUR   17,50
```

**Comment by AngledLuffa** [comment]
In terms of the CRF, currently no.  You could theoretically add it as a
feature to the model.  This doesn't really seem like a situation where you
want to use machine learning techniques to find the last row of a table,
though
